packages/blade/src/utils/useBreakpoint/useBreakpoint.ts

```diff
@@ -0,0 +1,144 @@
+              let mediaQuery = '';
+
+              if (breakpointTokenName === 'max') {
+                mediaQuery = `screen and (min-width: ${breakpointSize}px)`;
+              } else if (breakpointsArray[index - 1]) {
+                mediaQuery = `screen and (min-width: ${
+                  breakpointsArray[index - 1][1] + 1
+                }px) and (max-width: ${breakpointSize}px)`;
+              } else {
+                mediaQuery = `screen and (max-width: ${breakpointSize}px)`;
+              }
+
+              return [breakpointTokenName, breakpointSize, mediaQuery];
+            },
+          )
+        : []) as [keyof Breakpoints, number, string][],
+    [breakpoints, supportsMatchMedia],
+  );
+
+  const getMatchedDeviceType = useCallback((matchedBreakpoint: Breakpoint): DeviceType => {
+    let matchedDeviceType: DeviceType = 'mobile';
+    if (typeof navigator !== 'undefined' && navigator.product === 'ReactNative') {
+      // react-native
+      matchedDeviceType = 'mobile';
+    } else if (typeof document !== 'undefined') {
+      // browser
+      if (matchedBreakpoint && ['xs', 's', 'm'].includes(matchedBreakpoint)) {
+        // tablet is also categorised as mobile
+        matchedDeviceType = 'mobile';
+      } else {
+        matchedDeviceType = 'desktop';
+      }
+    } else if (typeof process !== 'undefined') {
+      // node
+      //@TODO: Check for useragent for node
+      matchedDeviceType = 'desktop';
+    }
+    return matchedDeviceType;
+  }, []);
+
+  const getMatchedBreakpoint = useCallback(
+    (event?: MediaQueryListEvent): Breakpoint => {
+      const matchedBreakpoint =
+        breakpointsTokenAndQueryCollection.find(([_, __, mediaQuery]) => {
+          if (event?.media === mediaQuery) {
+            return true;
+          }
+          if (window.matchMedia(mediaQuery).matches) {
+            return true;
+          }
+          return false;
+        })?.[0] ?? undefined;
+
+      return matchedBreakpoint;
+    },
+    [breakpointsTokenAndQueryCollection],
+  );
+
+  const [breakpointAndDevice, setBreakpointAndDevice] = useState(() => {
+    const matchedBreakpoint = getMatchedBreakpoint();
+    const matchedDeviceType = getMatchedDeviceType(matchedBreakpoint);
+    return {
+      matchedBreakpoint,
+      matchedDeviceType,
+    };
+  });
+
+  useEffect(() => {
+    if (!supportsMatchMedia) {
+      return undefined;
+    }
+
+    const handleMediaQueryChange = (event: MediaQueryListEvent): void => {
+      setBreakpointAndDevice((prevBreakpointAndDevice) => {
+        const matchedBreakpoint = getMatchedBreakpoint(event);
+        const matchedDeviceType = getMatchedDeviceType(matchedBreakpoint);
+        if (
+          prevBreakpointAndDevice.matchedBreakpoint !== matchedBreakpoint ||
+          prevBreakpointAndDevice.matchedDeviceType !== matchedDeviceType
+        ) {
+          return { matchedBreakpoint, matchedDeviceType };
+        }
+        return prevBreakpointAndDevice;
+      });
```

Comment:
this is the handler that gets triggered when the media query's `change` event is triggered. so we check if the `matchedBreakpoint` or `matchedDeviceType` has changed and update the state accordingly

Reply:
because it's setState so if I always return the new object it'll trigger unnecessary re-renders hence I return the old state if nothing changes. basically bail-out of re-render